pandas/tests/indexes/string/test_indexing.py

```diff
@@ -6,6 +6,37 @@
 import pandas._testing as tm
 
 
+class TestGetLoc:
+    def test_get_loc(self, any_string_dtype):
+        index = Index(["a", "b", "c"], dtype=any_string_dtype)
+        assert index.get_loc("b") == 1
+
+    def test_get_loc_raises(self, any_string_dtype):
+        index = Index(["a", "b", "c"], dtype=any_string_dtype)
+        with pytest.raises(KeyError, match="d"):
+            index.get_loc("d")
+
+    def test_get_loc_invalid_value(self, any_string_dtype):
+        index = Index(["a", "b", "c"], dtype=any_string_dtype)
+        with pytest.raises(KeyError, match="1"):
+            index.get_loc(1)
+
+    def test_get_loc_non_unique(self, any_string_dtype):
+        index = Index(["a", "b", "a"], dtype=any_string_dtype)
+        result = index.get_loc("a")
+        expected = np.array([True, False, True])
+        tm.assert_numpy_array_equal(result, expected)
+
+    def test_get_loc_non_missing(self, any_string_dtype, nulls_fixture):
+        index = Index(["a", "b", "c"], dtype=any_string_dtype)
+        with pytest.raises(KeyError):
+            index.get_loc(nulls_fixture)
+
+    def test_get_loc_missing(self, any_string_dtype, nulls_fixture):
```

Comment:
So this test now means that you can use `np.nan` and `pd.NA` interchangeably when indexing? If that's correct, I'm not sure I agree that we should be going that far

Reply:
The problem is that we are coercing any missing value indicator to NaN upon construction, and so to preserve back compat, I think I prefer we do the same for input to indexing operations.

To express it in terms of get_loc, this works now:

```python
>>> pd.options.future.infer_string = False
>>> pd.Index(["a", "b", None]).get_loc(None)
2
```

but the same on main with enabling the string dtype:

```python
>>> pd.options.future.infer_string = True
>>> pd.Index(["a", "b", None]).get_loc(None)
...
KeyError: None
```

That is because now the None is no longer in the object dtype index, but has been coerced to NaN.   
(on main, trying the above with `np.nan` also fails (see the issue https://github.com/pandas-dev/pandas/issues/59879), but that's because the `StringEngine` simply wasn't set up to work with missing values, so that is the initial reason I replaced it now with the StringObjectEngine)

The above is with `None`, but essentially happens with any other missing value indicator, like pd.NA. Maybe `None` and `np.nan` are the most important ones though, but I would at least prefer that indexing with `None` keeps working for now (we can always start deprecating it, but I wouldn't do that it as a breaking change for 3.0)

Reply:
FWIW this is also already quite inconsistent depending on the data type .. See https://github.com/pandas-dev/pandas/issues/59765 for an overview (e.g. also for datetimelike and categorical, we treat all NA-likes as the same in indexing lookups)

Reply:
> FWIW this is also already quite inconsistent depending on the data type .. See #59765 for an overview (e.g. also for datetimelike and categorical, we treat all NA-likes as the same in indexing lookups)

Nice - that's a great issue. Thanks for opening it.

> To express it in terms of get_loc, this works now:

Hmm I'm a bit confused by how this relates to all of the missing indicators becoming essentially equal though. On main, this does not work (?):

```python
>>> pd.options.future.infer_string = False
>>> pd.Index(["a", "b", None]).get_loc(np.nan)
KeyError: nan
```

Definitely understand that there is not an ideal solution here given the inconsistent history, but I don't want to go too far and just start making all of the missing value indicators interchangeable. I _think_ containment logic should land a little closer to equality logic, and in the latter we obviously don't allow this

Reply:
Fair point on the equality. I guess I'm still hung up on the indexing behavior being the same though.

I've lost track of the nuance a bit, but haven't np.nan and pd.NA always had different indexing behavior? I'm just wary of glossing over that as part of this.

Maybe worth some input from @pandas-dev/pandas-core if anyone else has thoughts

Reply:
I updated the PR to for now just enable exact matching missing values in `get_loc`, so this PR can already be merged (and fix the most glaring bug), and then we can have the discussion around backwards compatibility in https://github.com/pandas-dev/pandas/issues/59879 (I don't think the above thread is very easy to follow for other people to chime in, will do a write up on the issue -> see https://github.com/pandas-dev/pandas/issues/59879#issuecomment-2495525796)